src/libraries/System.Private.CoreLib/src/System/Boolean.cs

```diff
@@ -191,24 +187,14 @@ public int CompareTo(bool value)
 
         // Custom string compares for early application use by config switches, etc
         //
-        internal static bool IsTrueStringIgnoreCase(ReadOnlySpan<char> value)
-        {
-            return value.Length == 4 &&
-                    (value[0] == 't' || value[0] == 'T') &&
-                    (value[1] == 'r' || value[1] == 'R') &&
-                    (value[2] == 'u' || value[2] == 'U') &&
-                    (value[3] == 'e' || value[3] == 'E');
-        }
+        internal static bool IsTrueStringIgnoreCase(ReadOnlySpan<char> value) =>
+            value.Length == 4 &&
+            (BinaryPrimitives.ReadUInt64LittleEndian(MemoryMarshal.AsBytes(value)) | 0x0020002000200020) == 0x65007500720074; // "true" as a ulong, each char |'d with 0x0020 for case-insensitivity
```

Comment:
This could be just `value.StartsWith("true", OridnalIgnoreCase)` and optimized to exactly that ^
if https://github.com/dotnet/runtime/pull/46392 was merged (but nobody wanted to approve it :\)

Reply:
Unrolling for `String.StartsWith` https://github.com/EgorBo/runtime-1/commit/2c5c92e08405528b8577b07cae3e66edb99418d8 (unfortunately, inliner won't even bother scanning it currently - it's too big, but it works if I ask it to do so anyway):

![image](https://user-images.githubusercontent.com/523221/152525913-2fdc2a4e-9df4-4d7f-aed9-feadaac7fa87.png)

Reply:
`BinaryPrimitives.ReverseEndianness(((ulong)value[3] << 48) | ((ulong)value[2] << 32) | ((ulong)value[1] << 16) | value[0]);` is folded into a ulong constant as is 😄

Reply:
Are you going to reopen that PR?

Reply:
> Are you going to reopen that PR?

It needs to be done differently, e.g. when JIT sees `str.StartsWith("hello")` it should replace the call with some special internal method written in C# e.g. `StringHelpers.UnrolledStartsWith_up_to_32chars_OridnalIgnoreCase(str, "hello")` which is always inlineable and its implementation knows input is always constant.

It's way better than doing unrolling in JIT IR which is verbose, difficult to maintain and error-prone (but more throughput wise).

The only small issue that JIT is not able to inline such helper calls yet but it trivial to fix.
I can do it if I am guaranteed I won't waste time on it, e.g. @jkotas suggested source gens for that instead https://github.com/dotnet/runtime/issues/45613#issuecomment-739108650

In theory we can optimize `String`, `ReadOnlySpan<char>` and `ReadOnlyByte<byte>` (utf8) for `Equals/StartsWith/SequenceEquals/EndsWith` for `Ordinal/OrdinalIgnoreCase`

Reply:
E.g.:
```csharp
static bool UnrolledEquals_up_to_32_chars(string str, string value)
{
    if (value.Length == 1)
        // TODO: we can skip Length check for this case - we can take 2 bytes from Length and 2 bytes of '\0'
        return str.Length >= 1 && str._firstChar == value[0];

    if (value.Length == 2)
        return str.Length >= 2 && Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref str._firstChar)) ==
            BinaryPrimitives.ReverseEndianness(((uint)value[1] << 16) | value[0]);

    if (value.Length == 3)
        return str.Length >= 3 && Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref str._firstChar)) ==
            BinaryPrimitives.ReverseEndianness(((ulong)value[2] << 48) | ((ulong)value[1] << 32) | ((ulong)value[0] << 16));

    if (value.Length == 4)
        return str.Length >= 4 && Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref str._firstChar)) ==
            BinaryPrimitives.ReverseEndianness(((ulong)value[3] << 48) | ((ulong)value[2] << 32) | ((ulong)value[1] << 16) | value[0]);

    // 5..7 - via two ulongs (overlapped)

    // 8..16 - via two Vector128 (overlapped if needed)
    var mask = Vector128.Create(value[0], value[1], ...); // should be recognized as a constant vector

    // 16..32 - via two Vector256 (overlapped if needed)


    // we can split this long impl into 3 methods just to make jit's life easier if needed

    return str.Equals(value); // fallback just in case
}
```